### PR TITLE
Sku enhancements

### PR DESCRIPTION
**Pull request by MatteoPiovanelli-Laser**
This implements the first stage of enhancements discussed in #101.
Pasting from there:

>  sets up a feature that has:
>  
>   - A setting to enforce checks on the uniqueness of SKUs.
>   - A setting enabling automatic generation of SKUs:
>     - The generation works similarly to Autoroute, but it is based on a site setting, rather than a type setting.
>     - The default pattern is `"{SKU-{Content.Id}"`

With the difference that the default pattern used is `SKU-{Content.Slug}`

The settings for this go in a menu under `Commerce -> Settings`

### CODE REVIEW

File: Handlers/AdvancedSKUManagementHandler.cs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        protected override void UpdateEditorShape(UpdateEditorContext context) {
+            base.UpdateEditorShape(context);
+            //Apply this only for ProductParts
+            if (context.ContentItem.Parts.Any(part => part is ProductPart)) {
+                var productPart = context.ContentItem.As<ProductPart>();
+                var settings = _orchardServices.WorkContext.CurrentSite.As<AdvancedSKUsSiteSettingPart>();
+                if (settings.RequireUniqueSKU) {
+                    var sameSKUProductIds = _contentManager.Query<ProductPart, ProductPartRecord>(VersionOptions.Latest)
+                        .Where(ppr => ppr.Id != productPart.Record.Id && ppr.Sku == productPart.Sku)
+                        .List().Select(pp => pp.Id);
```
Comment (bleroy) [inline Handlers/AdvancedSKUManagementHandler.cs]: I think the `Select` should come before the `List()`.

File: Handlers/AdvancedSKUManagementHandler.cs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            if (string.IsNullOrWhiteSpace(part.Sku) ||
+                _SKUGenerationServices.Value.GetSettings().AllowCustomPattern) { //AllowCustomPattern is handled inside GenerateSku
+                //generate a new sku
+                part.Sku = _SKUGenerationServices.Value.GenerateSku(part);
+            }
+
+            //if the SKU is empty, generate a new one
+            if (string.IsNullOrWhiteSpace(part.Sku)) {
+                _SKUGenerationServices.Value.ProcessSku(part);
+                _orchardServices.Notifier.Warning(T("A new SKU has been generated: \"{0}\"", part.Sku));
```
Comment (bleroy) [inline Handlers/AdvancedSKUManagementHandler.cs]: This sounds very chatty, and worth an info message at best, not a warning.

File: Menus/ECommerceSettingsAdminMenu.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        public ECommerceSettingsAdminMenu() {
+            T = NullLocalizer.Instance;
+        }
+
+        public Localizer T { get; set; }
+
+        public void GetNavigation(NavigationBuilder builder) {
+            builder
+                .AddImageSet("nwazet-commerce")
+                .Add(item => item
```
Comment (bleroy) [inline Menus/ECommerceSettingsAdminMenu.cs]: I don't know that this should be under the commerce menu. I'd have expected to find a new commerce settings entry under the global settings header.

File: Services/LocalizationSKUUniquenessExceptionProvider.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+using System.Threading.Tasks;
+using Nwazet.Commerce.Models;
+using Orchard.ContentManagement;
+using Orchard.Environment.Extensions;
+using Orchard.Localization.Models;
+using Orchard.Localization.Services;
+
+namespace Nwazet.Commerce.Services {
+    [OrchardFeature("Nwazet.AdvancedSKUManagement")]
+    public class LocalizationSKUUniquenessExceptionProvider : ISKUUniquenessExceptionProvider {
```
Comment (bleroy) [inline Services/LocalizationSKUUniquenessExceptionProvider.cs]: Isn't that a helper more than an exception provider, since it doesn't seem to provide any exceptions?

File: ViewModels/AdancedSKUsSiteSettingsViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Orchard.Environment.Extensions;
+
+namespace Nwazet.Commerce.ViewModels {
+    [OrchardFeature("Nwazet.AdvancedSKUManagement")]
```
Comment (bleroy) [inline ViewModels/AdancedSKUsSiteSettingsViewModel.cs]: I don't think you need that attribute.

File: Views/EditorTemplates/Parts/AdvancedSKUProduct.cshtml [hunk 6]
```diff
@@ -0,0 +1,41 @@
+﻿@model Nwazet.Commerce.ViewModels.AdvancedSKUProductEditorViewModel
+
+@Html.HiddenFor(m => m.Product.Sku)
+
+@{
+    string skuHtml = "";
+}
+@if (Model.Settings.AllowCustomPattern) {
```
Comment (bleroy) [inline Views/EditorTemplates/Parts/AdvancedSKUProduct.cshtml]: All this concatenation and escaping is hard to follow and probably worse to maintain. I'd prefer to let Razor do its job (maybe in a hidden div), and have the script copy the results.

File: Handlers/AdvancedSKUManagementHandler.cs (hunk 1 above)
Comment (MatteoPiovanelli-Laser) [inline Handlers/AdvancedSKUManagementHandler.cs]: I would agree, but `IContentQuery` does not seem to have a `Select` method. Am I missing an `using` for that?

File: Handlers/AdvancedSKUManagementHandler.cs (hunk 2 above)
Comment (MatteoPiovanelli-Laser) [inline Handlers/AdvancedSKUManagementHandler.cs]: You are probably right, I will change that. Note however that in the corresponding code in the `ProcessAlias` method of `AutoroutePartHandler` it uses a warning.

File: Menus/ECommerceSettingsAdminMenu.cs (hunk 3 above)
Comment (MatteoPiovanelli-Laser) [inline Menus/ECommerceSettingsAdminMenu.cs]: That's where I started, but it felt more natural to have them together with everything else related to e-commerce. I feel it's better UX, but I can change it to the other setup if need be.

File: Services/LocalizationSKUUniquenessExceptionProvider.cs (hunk 4 above)
Comment (MatteoPiovanelli-Laser) [inline Services/LocalizationSKUUniquenessExceptionProvider.cs]: yes, you are right. They are not exceptions in the sense of c#'s Exceptions. I'll rename the interface (and related things).

File: ViewModels/AdancedSKUsSiteSettingsViewModel.cs (hunk 5 above)
Comment (MatteoPiovanelli-Laser) [inline ViewModels/AdancedSKUsSiteSettingsViewModel.cs]: No, I definetely don't

File: Views/EditorTemplates/Parts/AdvancedSKUProduct.cshtml (hunk 6 above)
Comment (MatteoPiovanelli-Laser) [inline Views/EditorTemplates/Parts/AdvancedSKUProduct.cshtml]: This is only one of the reasons why I should not be allowed to do the views of things. Anyway, fixing this.

File: Handlers/AdvancedSKUManagementHandler.cs (hunk 1 above)
Comment (bleroy) [inline Handlers/AdvancedSKUManagementHandler.cs]: You're right. If that ever becomes a bottleneck however, we can change that into more efficient HQL.

File: Menus/ECommerceSettingsAdminMenu.cs (hunk 3 above)
Comment (bleroy) [inline Menus/ECommerceSettingsAdminMenu.cs]: Yes, we should have what is managing content items (products, orders, shipping, etc.) under the commerce header, and everything that changes site settings under settings.

File: Views/EditorTemplates/Parts/AdvancedSKUProduct.cshtml [hunk 7]
```diff
@@ -4,36 +4,46 @@
...
-}
+</div>
+
 
 @using (Script.Foot()) {
     <script type="text/javascript">
         //replace the sku portion of the usual view with what we have here
         $(function () {
+            var myDiv = $("#@divId");
+            myDiv.remove();
```
Comment (bleroy) [inline Views/EditorTemplates/Parts/AdvancedSKUProduct.cshtml]: I'm nitpicking at this point, but I would have instead used an inline style to hide the div, instead of relying on script to remove it.

**Comment by bleroy** [comment]
Thanks for the changes. I'll let you decide if you want to address the new comments. In either case, I'll merge.

File: Views/EditorTemplates/Parts/AdvancedSKUProduct.cshtml (hunk 7 above)
Comment (MatteoPiovanelli-Laser) [inline Views/EditorTemplates/Parts/AdvancedSKUProduct.cshtml]: Done.

File: Menus/ECommerceSettingsAdminMenu.cs (hunk 3 above)
Comment (MatteoPiovanelli-Laser) [inline Menus/ECommerceSettingsAdminMenu.cs]: I will change this so the settings are under the global settings section, as is conventional in Orchard, but I nonetheless feel that this is worth discussing as a possible UX improvement.

**Comment by bleroy** [comment]
Thanks!
